### ISSUE TITLE
Renderer optimization tracking issue

### ISSUE THREAD

**Issue by superdump** [bevyengine/bevy#12590]
There are many ideas, branches, proofs of concepts, PRs, and discussions around improving the performance of the main code paths, systems, data structures, etc for rendering entities with meshes and materials.

This issue is a tracking issue to help with giving an overview of what has been considered, is known/has been tried, is almost ready but needs finishing off, needs reviewing, or has been merged. It should help to sequence work, and avoid forgetting things.

* [ ] Improved directional light shadow cascade culling
  * Cull to camera frustum slice bounds in light space instead of to the light frustum that covers camera frustum slice
* [ ] Faster `MeshUniform` serialization
  * Use manual padding and alignment, and `bytemuck` to bypass `encase`
* [ ] Consolidate `RenderAssets` and `RenderMaterials`
  * By making the generic type argument of `RenderAsset` be the target type (e.g. `GpuMesh` instead of `Mesh`) it is possible to make materials use `RenderAssets` infrastructure instead of having to have their own `RenderMaterials` solution, which has been copied from 3D to 2D, UI, gizmos...
  * This reduces any work on `RenderAssets` and `RenderMaterials` to just work on `RenderAssets`
* [ ] Slotmap `RenderAssets` (and `RenderMaterials`)
  * `AssetId` is quite large due to having an UUID variant (16 bytes) which means slower hashing and worse cache hit rates (larger data uses more space in caches). `Mesh` and `Material` `AssetId`s are stored in `PhaseItem`s which are sorted.
  * `HashMap` lookups hashing those `AssetId` keys is done a lot in the hottest paths (queuing, encoding draw commands to render passes).
  * `RenderAssets` could instead use a `SlotMap<PreparedAsset>`, and `HashMap<AssetId, SlotMapKey>`. On extraction, the `SlotMapKey` is looked up for the `AssetId`, and the render app then exclusively uses the `SlotMapKey` to look up the `PreparedAsset` which avoids hashing entirely, and means less data to be sorted.
* [ ] Split per-instance buffer dynamic offsets, and instance/batch ranges out of `PhaseItem`
  * Queuing, sorting, batching, rendering all interact with `RenderPhase` which contains a `Vec<PhaseItem>`
  * `PhaseItem` contains a `Range<u32>` (8 bytes) for the instance/batch range, and an `Option<NonMaxU32>` (4 bytes) for the dynamic offset in case of using a `BatchedUniformBuffer`. These take space in caches, and are more data to move around when sorting.
  * Queuing and sorting don't know anything about the instance/batch range nor dynamic offsets because they only get calculated in `batch_and_prepare_render_phase`
  * This leads to the conclusion that batch ranges and dynamic offsets should be moved out of `PhaseItem`s. They could be moved into separate `Vec` or `EntityHashMap` in `RenderPhase`, or separate components on the view, generic over the `PhaseItem` to allow there to be one per phase per view. The latter enables easier parallelism through ECS queries, but is perhaps a bit more awkward.
* [ ] Split `MeshUniform` `GpuArrayBuffer` preparation from `batch_and_prepare_render_phase`
  * `batch_and_prepare_render_phase` is a bottleneck
  * `batch_and_prepare_render_phase` prepares the per-instance data buffer because when using `BatchedUniformBuffer` on WebGL2/where storage buffers are not available, batches can be broken by filling up a uniform buffer binding (16kB minimum guaranteed, 64kB on NVIDIA, can be more on others) such that a new dynamic offset binding has to be started.
  * Dynamic offsets can be calculated without having to prepare the buffer

**Comment by Elabajaba** [comment]
Other things to add:

- [ ] Shadow atlasing for improved performance
  - This also allows users to set a fixed VRAM budget for shadows. Having a fixed budget means we also need to figure out how to gracefully and automatically degrade shadows based on distance (or some other metric) to fit everything into the atlas.
  - Caching this between frames and only updating shadow views when necessary (eg. don't update a view if it was static) can be a nice performance win.
- [ ] Reduce our register pressure in shaders
  - This is something that we'll eternally be fighting against, especially as more features are added.
  - Currently our main_opaque_3d fragment shader has insane register pressure due to how many VGPRs we're using, which means we get poor throughput. Forward shading is expected to have high register pressure, but bevy's is currently insanely high.
  - See https://gpuopen.com/learn/occupancy-explained/ for an explanation of GPU occupancy
  - See https://flashypixels.wordpress.com/2018/11/10/intro-to-gpu-scalarization-part-1/ and https://gpuopen.com/learn/optimizing-gpu-occupancy-resource-usage-large-thread-groups/ for some articles on how to improve occupancy.
  - [RGA](https://gpuopen.com/rga/) can be used if you dump bevy's shaders to spirv to view what's actually causing VGPR pressure, which is what's currently causing the low occupancy.
  - Nvidia Nsight and AMD RGP can show what the register pressure is in a capture. Intel GPA and Apple's Xcode gpu debugger can probably show this. Idk about android.
- [ ] FMA all the things.
  - Afaik shader compilers won't reorder floating point ops as it changes precision/output, so we have to try and write shaders so they generate fma.
  - Basically, make sure to write stuff as `a * b + c` instead of `c + a * b`

**Comment by james7132** [comment]
> Improve MeshUniform inverse matrix calculation performance 

This may be supplanted on platforms where compute shaders are present by #12773.

**Comment by pcwalton** [comment]
#12773 also effectively bypasses `encase` on platforms where compute shaders are present, since `MeshInputUniform` uses `bytemuck`.
